include/fi_util.h

```diff
@@ -103,6 +104,47 @@ enum {
 };
 
 /*
+ * Utility classes/functions
+ */
+/*
+ * The key_idx object and related functions can be used to generate unique keys
+ * from an index. The key and index would refer to an object defined by the user.
+ * A local endpoint can exchange this key with a remote endpoint in the first message.
+ * The remote endpoint would then use this key in subsequent messages to reference
+ * the correct object at the local endpoint.
+ */
+struct ofi_key_idx {
+	uint64_t seq_no;
+	/* The uniqueness of the generated key would depend on how many bits are
+	 * used for the index */
+	uint8_t idx_bits;
+};
+
+static inline struct ofi_key_idx *ofi_key_idx_alloc(size_t size)
+{
+	struct ofi_key_idx *key_idx = calloc(1, sizeof(*key_idx));
+	if (!key_idx)
+		return NULL;
+	key_idx->idx_bits = fi_size_bits((uint64_t)size);
+	return key_idx;
+}
```

Comment:
I would replace 'alloc' with 'init' and let the user just allocate the structure, since it only has 2 fields with no dynamic sizing.